src/gsi/core_disp_func.F90

```diff
@@ -0,0 +1,1055 @@
+!     num_glob_needed
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  implicit none
+  real(fp), dimension(:,:), intent(inout) :: x_disp, y_disp, dx_n, dy_n
+  integer, dimension(:,:), intent(inout) :: ix_orig, iy_orig
+  type (fca_gridded_meta), intent(in) :: meta		! meta data
+  integer, allocatable, intent(out) :: needed_ij_in(:,:), needed_ij_out(:,:), index_in(:)
+  real(fp), allocatable, intent(out) :: needed_wgts_out(:,:)
+  integer, intent(out) :: num_glob_needed
+  integer, intent(in) :: order
+  real(fp), allocatable :: needed_wgts_in(:,:)
+  integer :: num_needed_out
+
+  real(fp), dimension(size(x_disp,1),size(x_disp,2)) :: x, y
+  integer :: needed_ij_local(size(x_disp,1),size(x_disp,2),2)
+  integer :: i, j, status, num_needed_in, ierr, offset, buf, Nx, Ny
+  integer, parameter :: BAD_INT=-999999
+  integer :: i1, i2, j1, j2
+
+#ifdef TRACE_USE
+  if (trace_use) call da_trace_entry("compute_xy_orig_dm")
+#endif
+
+  Nx = size(x_disp,1)
+  Ny = size(x_disp,2)
+
+  forall(i=1:Nx,j=1:Ny) 
+     x(i,j) = its-1+(i-1)*meta%dx + x_disp(i,j) + meta%x0
+     y(i,j) = jts-1+(j-1)*meta%dy + y_disp(i,j) + meta%y0
+  end forall
+
+  ! new array indices.
+  ! the r code achieves this by performing integer division on two floats, which of course were implicitly cast.
+
+  ix_orig(1:Nx,1:Ny) = int(floor(x(1:Nx,1:Ny)/meta%dx)+1) ! x index of lower-left corner of containing value to bring back here
+  iy_orig(1:Nx,1:Ny) = int(floor(y(1:Nx,1:Ny)/meta%dy)+1) ! y ""
+
+  !	use constant extrapolation for points outside the grid:
+  if (order .eq. bilinear) then
+     buf=1
+  elseif (order .eq. bicubic) then
+     buf=2
+  end if
+  where(ix_orig < ids+buf-1)
+     dx_n = 0 
+     ix_orig = ids+buf-1
+  elsewhere (ix_orig > (ide-1-buf))
+     dx_n = 1
+     ix_orig = ide-1-buf
+  elsewhere
+     ! x weight for interpolating values from box (i.e. 0.25 -> 25% from bottom, 75% from top points)
+     dx_n = mod(x,meta%dx)/meta%dx
+  end where
+
+  where(iy_orig < jds+buf-1)
+     dy_n = 0
+     iy_orig = jds+buf-1
+  elsewhere (iy_orig > (jde-1-buf))
+     dy_n = 1
+     iy_orig = jde-1-buf
+  elsewhere
+     dy_n = mod(y,meta%dy)/meta%dy
+  end where
+
+  ! Build up a list of origin points and associated weights for which
+  ! interpolation will have to computed in another patch
+
+  ! prefill:
+  needed_ij_local(1:Nx,1:Ny,1:2)=BAD_INT
+  num_needed_in=0
+
+  ! Only apply displacements within same limits as in apply_disp_2d:
+
+  if (order .eq. bilinear) then
+     ! apply displacements up to and including domain boundary
+     j1=1
+     j2=Ny
+     i1=1
+     i2=Nx
+  elseif (order .eq. bicubic) then
+     ! apply displacements one gridpoint in from domain boundary
+     j1=max(jds+1,jts)-jts+1
+     j2=min(jde-1,jte)-jts+1
+     i1=max(ids+1,its)-its+1
+     i2=min(ide-1,ite)-its+1
+  end if
+    
+  do j = j1, j2
+     do i = i1, i2
+        if (ix_orig(i,j) .lt. ips .or. ix_orig(i,j) .gt. ipe .or. &
```

Comment:
Please replace .lt. with < and .gt. with >

Reply:
See above.